### ISSUE TITLE
Vulkan CTS test gets assert in Function::GetBlockDepth

### ISSUE THREAD

**Issue by Tony-LunarG** [KhronosGroup/SPIRV-Tools#517]
Vulkan CTS test dEQP-VK.spirv_assembly.instruction.graphics.selection_block_order.out_of_order_vert gets a "vector subscript out of range" assert when run with validation layers on Windows.  The stack trace points to PerformCfgChecks calling Function::GetBlockDepth then at line 402 continue_construct->corresponding_constructs is empty and the vector code asserts.  This started happening with commit 6993fc4 "Validation code for control flow nesting depth".

**Comment by ehsannas** [comment]
I'll take a look tonight or tomorrow.

**Comment by ehsannas** [comment]
![1](https://cloud.githubusercontent.com/assets/3828215/21686677/3ec1de36-d334-11e6-853d-a1162b1868e1.png)

This is a representation of the control flow graph in of the failing test. 

The issue is caused because `74` is both the entry block of the continue construct and also the entry block of the selection construct.

Here's the SPIRV code for this test:
```
; SPIR-V
; Version: 1.0
; Generator: Khronos SPIR-V Tools Assembler; 0
; Bound: 84
; Schema: 0
               OpCapability Shader
               OpCapability ClipDistance
               OpCapability CullDistance
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main" %_ %position %vtxColor %color %gl_VertexIndex %gl_InstanceIndex
               OpName %main "main"
               OpName %gl_PerVertex "gl_PerVertex"
               OpMemberName %gl_PerVertex 0 "gl_Position"
               OpMemberName %gl_PerVertex 1 "gl_PointSize"
               OpMemberName %gl_PerVertex 2 "gl_ClipDistance"
               OpMemberName %gl_PerVertex 3 "gl_CullDistance"
               OpName %testfun_vf4_ "testfun(vf4;"
               OpName %_ ""
               OpName %position "position"
               OpName %vtxColor "vtxColor"
               OpName %color "color"
               OpName %gl_VertexIndex "gl_VertexIndex"
               OpName %gl_InstanceIndex "gl_InstanceIndex"
               OpMemberDecorate %gl_PerVertex 0 BuiltIn Position
               OpMemberDecorate %gl_PerVertex 1 BuiltIn PointSize
               OpMemberDecorate %gl_PerVertex 2 BuiltIn ClipDistance
               OpMemberDecorate %gl_PerVertex 3 BuiltIn CullDistance
               OpDecorate %gl_PerVertex Block
               OpDecorate %position Location 0
               OpDecorate %vtxColor Location 1
               OpDecorate %color Location 1
               OpDecorate %gl_VertexIndex BuiltIn VertexIndex
               OpDecorate %gl_InstanceIndex BuiltIn InstanceIndex
       %void = OpTypeVoid
       %bool = OpTypeBool
        %int = OpTypeInt 32 1
       %uint = OpTypeInt 32 0
      %float = OpTypeFloat 32
    %v3float = OpTypeVector %float 3
    %v4float = OpTypeVector %float 4
     %v4bool = OpTypeVector %bool 4
         %18 = OpTypeFunction %v4float %v4float
         %19 = OpTypeFunction %void
%_ptr_Input_float = OpTypePointer Input %float
%_ptr_Input_int = OpTypePointer Input %int
%_ptr_Input_v3float = OpTypePointer Input %v3float
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_float = OpTypePointer Output %float
%_ptr_Output_v4float = OpTypePointer Output %v4float
%_ptr_Function_float = OpTypePointer Function %float
%_ptr_Function_int = OpTypePointer Function %int
%_ptr_Function_v4float = OpTypePointer Function %v4float
    %float_1 = OpConstant %float 1
    %float_0 = OpConstant %float 0
  %float_0_5 = OpConstant %float 0.5
   %float_n1 = OpConstant %float -1
    %float_7 = OpConstant %float 7
    %float_8 = OpConstant %float 8
      %int_0 = OpConstant %int 0
      %int_1 = OpConstant %int 1
      %int_2 = OpConstant %int 2
      %int_3 = OpConstant %int 3
      %int_4 = OpConstant %int 4
     %uint_0 = OpConstant %uint 0
     %uint_1 = OpConstant %uint 1
     %uint_2 = OpConstant %uint 2
     %uint_3 = OpConstant %uint 3
    %uint_32 = OpConstant %uint 32
     %uint_4 = OpConstant %uint 4
%uint_2147483647 = OpConstant %uint 2147483647
         %47 = OpConstantComposite %v4float %float_1 %float_1 %float_1 %float_1
         %48 = OpConstantComposite %v4float %float_1 %float_0 %float_0 %float_1
         %49 = OpConstantComposite %v4float %float_0_5 %float_0_5 %float_0_5 %float_0_5
%_arr_float_uint_1 = OpTypeArray %float %uint_1
%_arr_float_uint_2 = OpTypeArray %float %uint_2
%_arr_v4float_uint_3 = OpTypeArray %v4float %uint_3
%_arr_float_uint_4 = OpTypeArray %float %uint_4
%_arr_v4float_uint_32 = OpTypeArray %v4float %uint_32
%_ptr_Input__arr_v4float_uint_3 = OpTypePointer Input %_arr_v4float_uint_3
%_ptr_Input__arr_v4float_uint_32 = OpTypePointer Input %_arr_v4float_uint_32
%_ptr_Output__arr_float_uint_2 = OpTypePointer Output %_arr_float_uint_2
%_ptr_Output__arr_v4float_uint_3 = OpTypePointer Output %_arr_v4float_uint_3
%_ptr_Output__arr_float_uint_4 = OpTypePointer Output %_arr_float_uint_4
%gl_PerVertex = OpTypeStruct %v4float %float %_arr_float_uint_1 %_arr_float_uint_1
%_ptr_Output_gl_PerVertex = OpTypePointer Output %gl_PerVertex
          %_ = OpVariable %_ptr_Output_gl_PerVertex Output
   %position = OpVariable %_ptr_Input_v4float Input
   %vtxColor = OpVariable %_ptr_Output_v4float Output
      %color = OpVariable %_ptr_Input_v4float Input
%gl_VertexIndex = OpVariable %_ptr_Input_int Input
%gl_InstanceIndex = OpVariable %_ptr_Input_int Input
       %main = OpFunction %void None %19
         %61 = OpLabel
         %62 = OpLoad %v4float %position
         %63 = OpAccessChain %_ptr_Output_v4float %_ %int_0
               OpStore %63 %62
         %64 = OpLoad %v4float %color
         %65 = OpFunctionCall %v4float %testfun_vf4_ %64
               OpStore %vtxColor %65
               OpReturn
               OpFunctionEnd
%testfun_vf4_ = OpFunction %v4float None %18
         %66 = OpFunctionParameter %v4float
         %67 = OpLabel
         %68 = OpVariable %_ptr_Function_int Function
         %69 = OpVariable %_ptr_Function_v4float Function
               OpStore %68 %int_0
               OpStore %69 %66
               OpBranch %70
         %70 = OpLabel
         %71 = OpLoad %int %68
         %72 = OpSLessThan %bool %71 %int_4
               OpLoopMerge %73 %74 None
               OpBranchConditional %72 %74 %73
         %73 = OpLabel
         %75 = OpLoad %v4float %69
               OpReturnValue %75
         %74 = OpLabel
         %76 = OpAccessChain %_ptr_Function_float %69 %71
         %77 = OpIEqual %bool %71 %int_0
               OpSelectionMerge %78 None
               OpBranchConditional %77 %79 %80
         %80 = OpLabel
         %81 = OpLoad %float %76
         %82 = OpFSub %float %float_1 %81
               OpStore %76 %82
               OpBranch %78
         %78 = OpLabel
         %83 = OpIAdd %int %71 %int_1
               OpStore %68 %83
               OpBranch %70
         %79 = OpLabel
               OpStore %76 %float_0
               OpBranch %78
               OpFunctionEnd
```

**Comment by ehsannas** [comment]
I'll investigate further.

**Comment by ehsannas** [comment]
`entry_block_to_construct_` maps an entry block to its construct. The key in this map (the entry block) is not unique, and therefore the entry for the continue construct gets overwritten when the selection construct is discovered.

**Comment by ehsannas** [comment]
Although this is a bug in `entry_block_to_construct_`, this has not caused any issues because the only usage of it has been safe. The only real usage of the map is in the `FindConstructForEntryBlock` function. The only usage of this function is in function.cpp:216:
```
  if (current_block_->is_type(kBlockTypeLoop)) {
    ......
    auto continue_target =
        FindConstructForEntryBlock(current_block_)
            .corresponding_constructs()
            .back()
            ->entry_block();
    ......
  }
```
In this case, `FindConstructForEntryBlock` is called when the block type is `kBlockTypeLoop` which is a case in which a conflict doesn't occur.

**Comment by ehsannas** [comment]
I am working on a fix. Since a given block may be the entry block of different types of constructs, the `(basic_block, construct_type)` pair should be able to uniquely identify the construct.

**Comment by ehsannas** [comment]
This is fixed via 109ce225c9f7538ab56d52f036a08f16a15327a0
